tools/adaparse.cpp

```diff
@@ -1,8 +1,169 @@
 #include <cstdlib>
+#include <chrono>
 #include <iostream>
 #include <string_view>
+#include <sys/stat.h>
+#include <fmt/core.h>
 #include <cxxopts.hpp>
+#include <fstream>
+#include <fmt/os.h>
+#include <cerrno>
+
+#include <fmt/os.h>
+#include <typeinfo>
+#include <iostream>
+#include <cxxabi.h>
+#include <memory>
+
 #include "ada.h"
+#include "line_iterator.h"
+#ifdef _MSC_VER
+#include <io.h>
+#else
+#include <unistd.h>
+#endif
+
+uint64_t nano() {
+  return std::chrono::duration_cast<::std::chrono::nanoseconds>(
+             std::chrono::steady_clock::now().time_since_epoch())
+      .count();
+}
+
+template <typename Callable>
+bool print_part(Callable&& fmt_or_adaparse_print, std::string_view get_part,
+                const ada::url_aggregator& url) {
+  if (get_part.size() == 4) {
```

Comment:
Not a blocker, but checking size seems like an optimization for performance. I don't think we really need it, to make the implementation as readable as possible but, I don't have strong feelings about it.